webapp/src/components/expanded_view/reaction_button.tsx

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {forwardRef, RefObject, useImperativeHandle, useRef, useState} from 'react';
+import Picker from '@emoji-mart/react';
+
+import styled, {css} from 'styled-components';
+
+import {OverlayTrigger} from 'react-bootstrap';
+
+import ControlsButton from 'src/components/expanded_view/controls_button';
+import {MAKE_REACTION, RAISE_LOWER_HAND, reverseKeyMappings} from 'src/shortcuts';
+import SmileyIcon from 'src/components/icons/smiley_icon';
+import UnraisedHandIcon from 'src/components/icons/unraised_hand';
+import RaisedHandIcon from 'src/components/icons/raised_hand';
+import * as Telemetry from 'src/types/telemetry';
+import {StyledTooltip} from 'src/components/shared';
+import Shortcut from 'src/components/shortcut';
+import CompassIcon from 'src/components/icons/compassIcon';
+import {Emoji} from 'src/components/emoji/emoji';
+import {EmojiData} from 'src/types/types';
+
+const EMOJI_VERSION = '13';
+const EMOJI_SKINTONE_MAP = new Map([[1, ''], [2, '1F3FB'], [3, '1F3FC'], [4, '1F3FD'], [5, '1F3FE'], [6, '1F3FF']]);
+
+interface EmojiPickEvent {
+    id: string;
+    keywords?: string[]
+    name?: string;
+    native?: string;
+    shortcodes?: string;
+    skin?: number;
+    unified: string;
+}
+
+const getCallsClient = () => {
+    return window.opener ? window.opener.callsClient : window.callsClient;
+};
+
+export type ReactionButtonRef = {
+    toggle: () => void,
+};
+
+interface Props {
+    trackEvent: (event: Telemetry.Event, source: Telemetry.Source, props?: Record<string, any>) => void,
+}
+
+export const ReactionButton = forwardRef(({trackEvent}: Props, ref) => {
+    const barRef: RefObject<HTMLDivElement> = useRef<HTMLDivElement>(null);
+    const [showPicker, setShowPicker] = useState(false);
+    const [showBar, setShowBar] = useState(false);
+    useImperativeHandle(ref, () => ({
+        toggle() {
+            toggleReactions();
+        },
+    }));
```

Comment:
Is this actually needed? Mostly for my own understanding but couldn't we simply let the parent provide the handler?

Reply:
I have a branch where I've made `ReactionButton` a controlled component. The problem with that is that the expanded_view component is a class component, and any change to its state object forces a complete rerender. With this way, the only component that rerenders is the ReactionButton.
When we convert the expanded_view component to a functional hooks component we can get rid of this. :)